### DISCUSSION TITLE
CPU Pinning For KVM and Guest VMs

### DISCUSSION THREAD

**Discussion by btzq** [apache/cloudstack#9801] [Q&A] [answered]
Hi Community, 

We are thinking if its possible for:

Idea 1: CPU Pin specific cores of the Physical Server, for KVM Hypervisor purposes only.
- Goal of this is to avoid context switching between running the hypervisor and guest VMs

Idea 2: CPU Pin specific vCPU to guest VMs
- Goal is to improve performance of guest VMs, especially in latency sensitive applications (stock trading). 
- Live Migration is important to us, will this feature break it?

Has anyone achieved this before? How did you do it?

**Comment by rajujith** [comment]
@btzq  regarding 'Idea 1' have you tried extraconfig or agent hooks? We will need some testing to see if live migration could break.

**Reply by btzq** [reply]
@rajujith no we havent tried those out. Just wondering if someone else has done this before so we could learn from them. 

When you meant agent hooks, do you mean KVM Agent Hooks?
https://cwiki.apache.org/confluence/display/CLOUDSTACK/KVM+hook+script+include

**Reply by rajujith** [reply]
@btzq  I can think of three ways, but I haven't tried these with CPU pinning. Let's see if anyone has tried CPU pinning. 
1) Extraconfig : https://www.shapeblue.com/cloudstack-feature-first-look-enable-sending-of-arbitrary-configuration-data-to-vms/
2) KVM Agent hooks: https://github.com/apache/cloudstack/pull/3839
3) Libvirt hooks: https://docs.cloudstack.apache.org/en/latest/adminguide/hosts.html#kvm-libvirt-hook-script-include

**Comment by weizhouapache** [comment, marked as answer]
You can try libvirt hooks which can change the VM xml definition before/after vm is started/stopped/migrated.
